pallets/deitos/src/lib.rs

```diff
@@ -113,58 +135,177 @@ pub mod pallet {
     #[pallet::generate_deposit(pub(super) fn deposit_event)]
     pub enum Event<T: Config> {
         /// A user has successfully set a new value.
-        SomethingStored {
-            /// The new value set.
-            something: u32,
-            /// The account who set the new value.
-            who: T::AccountId,
+        IPRegistered {
+            ip: T::AccountId,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        },
+        IPStorageUpdated {
+            ip: T::AccountId,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        },
+        IPStatusChanged {
+            ip: T::AccountId,
+            status: IPStatus,
+        },
+        IPUnregistered {
+            ip: T::AccountId,
         },
     }
 
     /// information.
     #[pallet::error]
     pub enum Error<T> {
         /// The value retrieved was `None` as no value was previously set.
-        NoneValue,
-        StorageOverflow,
         NonExistentStorageValue,
+        /// Math overflow
+        Overflow,
+        /// Insufficient storage
+        InsufficientStorage,
+        /// On going agreements
+        OnGoingAgreements,
     }
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         #[pallet::call_index(0)]
         #[pallet::weight(T::WeightInfo::register_ip())]
         /* to add more parameters*/
-        pub fn register_ip(origin: OriginFor<T>, price_unit: BalanceOf<T>) -> DispatchResult {
+        pub fn register_ip(
+            origin: OriginFor<T>,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        ) -> DispatchResult {
             // Check that the extrinsic was signed and get the signer.
-            let who = ensure_signed(origin)?;
+            let ip = ensure_signed(origin)?;
+            let deposit_amount = IPDepositAmount::<T>::get()?;
+
+            T::Currency::hold(&HoldReason::IPInitialDeposit.into(), &ip, deposit_amount)?;
+
+            let ip_details = IPDetails::<T> {
+                price_storage_per_block,
+                total_storage,
+                reserved_storage: Zero::zero(),
+                status: IPStatus::Validating,
+                active_agreements: BoundedVec::new(),
+            };
+
+            InfrastructureProvider::<T>::insert(&ip, ip_details);
```

Comment:
I think we should insert only if IP is not there.

Reply:
https://github.com/Deitos-Network/deitos-node/pull/6#discussion_r1422246532